sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalFunctions.scala

```diff
@@ -28,6 +29,15 @@ case class UnscaledValue(child: Expression) extends UnaryExpression {
   override def nullable: Boolean = child.nullable
   override def toString: String = s"UnscaledValue($child)"
 
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (child.dataType.isInstanceOf[DecimalType]) {
+      TypeCheckResult.TypeCheckSuccess
+    } else {
+      TypeCheckResult.TypeCheckFailure(
+        s"UnscaledValue only accept decimal type, not ${child.dataType}")
```

Comment:
This thing is only added by the optimizer so I guess it is not that big of a deal to do this check... can you add an inline comment for this expression saying this is internal and created only by the optimizer?